### ISSUE TITLE
Derivation time stamps updated to incorrect values

### ISSUE THREAD

**Issue by sm453** [cambridge-cares/TheWorldAvatar#99]
The value of the time stamp at which a derivation was last updated must be the instant in time at which the inputs to the derivation (i.e. the entities which the derivation is derived from) were read. The value of the time stamp at which a derivation was last updated MUST NOT be the instant in time the entities belonging to the derivation (i.e. the entities whose values are derived) were written.

Unfortunately, the current implementation appears to do the latter, not the former. This must be rectified.

Please note that the difference between the two implementations is critical. In the latter one, we are susceptible to the nasty bug that an input value was updated AFTER it was read but BEFORE the derivation was updated, but the time stamp associated with the derivation will be AFTER the update of the input, thus erroneously suggesting the derivation is up to date while in reality it is not!

One may argue that, for synchronous derivations, this is not too serious because the wall time delay between reading an input and updating a derivation may be 'small'. For asynchronous derivations, this is much more serious, because by definition the time delay between reading the inputs and writing the outputs is 'large'.

I am therefore convinced that, for a robust implementation, both modes (synchronous and asynchronous) must update a derivation time stamp to the instant in time at which the inputs were READ.

**Comment by kokfoongcmcl** [comment]
The values are read by the agent itself, and the derivation client only provides the IRIs to the agent. For the synchronous side, the easiest thing I can do is to change the timestamp to the timestamp at the point when it sends the HTTP request to the agent.

I think what you really want is for the agent to update the timestamp at the point it reads the input. To do this really properly and efficiently on the synchronous side, the agent needs to include the timestamp in its HTTP response. This is because the timestamps are updated in 1 go only after all the derivations are updated (updating the timestamps 1 by 1 after each agent call is inefficient). Do you prefer this option? This enforces another kind of manual requirement on the agent that is difficult to enforce with a superclass.

There is a public method in the derivation client that updates the timestamp for a given derivation IRI or the instances belong to a derivation, I guess the asynchronous agents can use this method.

**Comment by kokfoongcmcl** [comment]
Actually we can create three classes: 1) derivation agent, 2) agent inputs, 3) agent outputs. The "derivation agent" class will have "agent inputs" as its argument and must return an "agent outputs" object. This "agent outputs" object will have timestamp as the compulsory argument in the constructor.

The "derivation agent" class will wrap around the HttpServlet class, serialises and deserialises the "agent inputs" and "agent outputs" objects (from/to JSON) behind the scenes. This way developers will not have to construct JSON objects manually themselves.

**Comment by jb2197** [comment]
For asynchronous operation, the inputs of a derivation are read by (asynchronous) agent when it detects the status of the derivation being marked as "Requested". The timestamp of the asynchronous derivation is updated when the agent detects the status of the derivation being marked as "Finished". I agree that the timestamp of derivation should be updated to the time instant when the inputs were read, and such an update should still be done at "Finished" status. Therefore, I think we can achieve it via introducing a new data property (for example, "retrievedInputsAt") to record the numeric position of unix time when the agent reading the inputs. When changing status at "Requested", we will add:
```
<derivation> <OntoDerivation:hasStatus> <status> .
<status> a <OntoDerivation:Requested> .
<status> <OntoDerivation:retrievedInputsAt> 123 .
```
the numeric position can be preserved until the end of the update and we will have something like:
```
<derivation> <OntoDerivation:hasStatus> <status> .
<status> a <OntoDerivation:Finished> .
<status> <OntoDerivation:retrievedInputsAt> 123 .
<status> <OntoDerivation:hasNewDerivedIRI> <derivedIRI> .
```
the last two lines can then be used during cleaning up.

> There is a public method in the derivation client that updates the timestamp for a given derivation IRI or the instances belong to a derivation, I guess the asynchronous agents can use this method.

Regarding the method to update the timestamp, I am currently using https://github.com/cambridge-cares/TheWorldAvatar/blob/develop/JPS_BASE_LIB/src/main/java/uk/ac/cam/cares/jps/base/derivation/DerivationSparql.java#L1243. I just noticed https://github.com/cambridge-cares/TheWorldAvatar/blob/develop/JPS_BASE_LIB/src/main/java/uk/ac/cam/cares/jps/base/derivation/DerivationSparql.java#L1659 and will change to the latter once we agreed the way forward.

**Comment by jb2197** [comment]
@kokfoongcmcl and I have discussed and agreed on the following:

### short term
1. for synchronous operation, the timestamp will be changed to the point when DerivationClient sends the HTTP request
2. for asynchronous operation, the timestamp will be changed to the point when the agent set up the job and will be written to KG with `{<status> <OntoDerivation:retrievedInputsAt> 123}`, and this recorded value will be used for updating the real timestamp once the status of asynchronous derivation is marked as "Finished"

### long term
> Actually we can create three classes: 1) derivation agent, 2) agent inputs, 3) agent outputs. The "derivation agent" class will have "agent inputs" as its argument and must return an "agent outputs" object. This "agent outputs" object will have timestamp as the compulsory argument in the constructor.
> 
> The "derivation agent" class will wrap around the HttpServlet class, serialises and deserialises the "agent inputs" and "agent outputs" objects (from/to JSON) behind the scenes. This way developers will not have to construct JSON objects manually themselves.

we will develop three classes as suggested with the following modifications: Java AsynAgent will be made part of "derivation agent" - we will have one agent class that can be extended into both synchronous and asynchronous agents, this will be the first step to have one framework supporting both modes
```
public class DerivationAgent extends HttpServlet implements DerivationAgentInterface {

    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
        JSONObject request_json = processHttpServletRequest(request);
        if validateInput(request_json) {
            // serialises DerivationInputs objects from JSONObject
            DerivationInputs inputs = new DerivationInputs(request_json);
            // deserialises DerivationOutputs objects to JSONObject as HTTP response
            DerivationOutputs outputs = processRequestParameters(inputs);
            response.setEntity(outputs.toJson());
        }
    }

    // agents override this
    public DerivationOutputs processRequestParameters(DerivationInputs derInputs) {
        // process the DerivationInputs objects
        // (it can either from the HTTP request)
        // (or the inputs from KG as part of asynchronous derivation)
        // generate DerivationOutputs objects with IRIs of new created instances
    }
    
    private JSONObject processHttpServletRequest(HttpServletRequest request) {
        // extract JSON object from request
    }

    public boolean validateInput(JSONObject requestParams) {
        // validate the inputs
    }

    public void monitorDerivation(String agentIRI) {
        // similar function as defined in AsynAgent.java
        // use processRequestParameters to setup job
    }
}
```

One example synchronous agent will look like:
```
public class OneSyncAgent extends DerivationAgent {

    @Override
    public boolean validateInput(JSONObject requestParams) {
        
    }

    @Override
    public DerivationOutputs processRequestParameters(DerivationInputs derInputs) {
        
    }
}
```

One example asynchronous agent will look like:
```
public class OneAsyncAgent extends DerivationAgent {

    @Override
    public boolean validateInput(JSONObject requestParams) {
        
    }

    @Override
    public DerivationOutputs processRequestParameters(DerivationInputs derInputs) {

    }

    @Override
    public void init() {
        // use ScheduledExecutorService to schedule execution of method monitorDerivation
    }
}
```

**Comment by sm453** [comment]
The short term items 1 and 2 have been addressed in pull requests #102 and #103, respectively. This (largely) fixes the issue, I believe.
Improving the code in the longer term so that we achieve a single unified framework supporting both modes is a separate topic.
